doc/source/reference/c-api/array.rst

```diff
@@ -1555,8 +1555,296 @@ For all of these macros *arr* must be an instance of a (subclass of)
+         identity otherwise as it preserves the sign for ``sum([-0.0])``.
+       - We use no identity for object, but return the default of ``0`` and
+         ``1`` for the empty ``sum([], dtype=object)`` and
+         ``prod([], dtype=object)``.
+         This allows ``np.sum(np.array(["a", "b"], dtype=object))`` to work.
+       - ``-inf`` or ``INT_MIN`` for ``max`` is an identity, but at least
+         ``INT_MIN`` not a good *default* when there are no items.
+
+   *initial* is a pointer to the data for the initial value, which should be
+   filled in. Returns -1, 0, or 1 indicating error, no initial value, and the
+   initial value being successfully filled. Errors must not be given when no
+   initial value is correct, since NumPy may call this even when it is not
+   strictly necessary to do so.
+
+.. c:type:: int (PyArrayMethod_TraverseLoop)( \
+        void *traverse_context, PyArray_Descr *descr, char *data, \
+        npy_intp size, npy_intp stride, NpyAuxData *auxdata)
+
+   A traverse loop working on a single array. This is similar to the general
+   strided-loop function. This is designed for loops that need to visit every
+   element of a single array.
+
+   Currently this is used for array clearing, via the ``NPY_DT_get_clear_loop``
+   DType API hook, and zero-filling, via the ``NPY_DT_get_fill_zero_loop``
+   DType API hook.  These are most useful for handling arrays storing embedded
+   references to python objects or heap-allocated data.
+
+   The *descr* is the descriptor for the array, *data* is a pointer to the array
+   buffer, *size* is the 1D size of the array buffer, *stride* is the stride,
+   and *auxdata* is optional extra data for the loop.
+
+   The *traverse_context* is passed in because we may need to pass in
+   Interpreter state or similar in the future, but we don't want to pass in a
+   full context (with pointers to dtypes, method, caller which all make no sense
+   for a traverse function). We assume for now that this context can be just
+   passed through in the the future (for structured dtypes).
+
+.. c:type:: int (PyArrayMethod_GetTraverseLoop)( \
+                void *traverse_context, PyArray_Descr *descr, \
+                int aligned, npy_intp fixed_stride, \
+                PyArrayMethod_TraverseLoop **out_loop, NpyAuxData **out_auxdata, \
+                NPY_ARRAYMETHOD_FLAGS *flags)
+
+   Simplified get_loop function specific to dtype traversal
+
+   It should set the flags needed for the traversal loop and set *out_loop* to the
+   loop function, which must be a valid ``PyArrayMethod_TraverseLoop``
+   pointer. Currently this is used for zero-filling and clearing arrays storing
+   embedded references.
+
+.. c:type:: int (PyArrayMethod_PromoterFunction)(PyObject *ufunc, \
+                PyArray_DTypeMeta *op_dtypes[], PyArray_DTypeMeta *signature[], \
+                PyArray_DTypeMeta *new_op_dtypes[])
+
+   Type of the promoter function, which must be wrapped into a
```

Comment:
Would be nice if, like for the other functions, this tells what the function should do! Probably can just adjust the text from below:
```
It is passed the operation and requested DType signatures and can mutate it to
attempt a new search for a matching loop/promoter.
```